ext/drmgr/drmgr.c

```diff
@@ -992,18 +973,79 @@ drmgr_bb_event(void *drcontext, void *tag, instrlist_t *bb, bool for_trace,
 
     pt->cur_phase = DRMGR_PHASE_NONE;
 
+    /* Delete table only if opcode events were ever registered by the user. */
+    if (local_info->was_opcode_instrum_registered)
+        hashtable_delete(&local_opcode_instrum_table);
+
+    return res;
+}
+
+static void
+drmgr_bb_event_set_local_cb_info(void *drcontext, OUT local_cb_info_t *local_info)
+{
+    dr_rwlock_read_lock(bb_cb_lock);
+    /* We use arrays to more easily support unregistering while in an event (i#1356).
+     * With arrays we can make a temporary copy and avoid holding a lock while
+     * delivering events.
+     */
+    cblist_create_local(drcontext, &cblist_app2app, &local_info->iter_app2app,
+                        (byte *)local_info->app2app,
+                        BUFFER_SIZE_ELEMENTS(local_info->app2app));
+    cblist_create_local(drcontext, &cblist_instrumentation, &local_info->iter_insert,
+                        (byte *)local_info->insert,
+                        BUFFER_SIZE_ELEMENTS(local_info->insert));
+    cblist_create_local(drcontext, &cblist_instru2instru, &local_info->iter_instru,
+                        (byte *)local_info->instru,
+                        BUFFER_SIZE_ELEMENTS(local_info->instru));
+    local_info->was_opcode_instrum_registered = was_opcode_instrum_registered;
+    /* We do not make a complete local copy of the opcode hashtable as this can be
+     * expensive. Instead, we create a scoped table later on that only maps the cb lists
+     * of those opcodes required by this specific bb.
+     */
+    dr_rwlock_read_unlock(bb_cb_lock);
+}
+
+static void
+drmgr_bb_event_delete_local_cb_info(void *drcontext, IN local_cb_info_t *local_info)
+{
+
+    cblist_delete_local(drcontext, &local_info->iter_app2app,
+                        BUFFER_SIZE_ELEMENTS(local_info->app2app));
+    cblist_delete_local(drcontext, &local_info->iter_insert,
+                        BUFFER_SIZE_ELEMENTS(local_info->insert));
+    cblist_delete_local(drcontext, &local_info->iter_instru,
+                        BUFFER_SIZE_ELEMENTS(local_info->instru));
+}
+
+static dr_emit_flags_t
+drmgr_bb_event(void *drcontext, void *tag, instrlist_t *bb, bool for_trace,
+               bool translating)
+{
+    dr_emit_flags_t res;
+    local_cb_info_t local_info;
+    void **pair_data = NULL, **quartet_data = NULL;
+    per_thread_t *pt = (per_thread_t *)drmgr_get_tls_field(drcontext, our_tls_idx);
+
+    drmgr_bb_event_set_local_cb_info(drcontext, &local_info);
+
+    /* We need per-thread user_data */
+    if (pair_count > 0)
```

Comment:
This needs to be merged with the race fix -- so this needs to be moved inside the _set_ function with the lock and added to the local struct or sthg right?

Reply:
Yep, done

Reply:
I tried to break down the PRs for easier reviews. Hope it didn't yield confusion instead.

Reply:
No, small is good.